// @nktkas/hyperliquid/script/src/api/info/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoClient = void 0;
// =============================================================
// Methods Imports
// =============================================================
const activeAssetData_js_1 = require("./_methods/activeAssetData.js");
const alignedQuoteTokenInfo_js_1 = require("./_methods/alignedQuoteTokenInfo.js");
const allMids_js_1 = require("./_methods/allMids.js");
const allPerpMetas_js_1 = require("./_methods/allPerpMetas.js");
const blockDetails_js_1 = require("./_methods/blockDetails.js");
const candleSnapshot_js_1 = require("./_methods/candleSnapshot.js");
const clearinghouseState_js_1 = require("./_methods/clearinghouseState.js");
const delegations_js_1 = require("./_methods/delegations.js");
const delegatorHistory_js_1 = require("./_methods/delegatorHistory.js");
const delegatorRewards_js_1 = require("./_methods/delegatorRewards.js");
const delegatorSummary_js_1 = require("./_methods/delegatorSummary.js");
const exchangeStatus_js_1 = require("./_methods/exchangeStatus.js");
const extraAgents_js_1 = require("./_methods/extraAgents.js");
const frontendOpenOrders_js_1 = require("./_methods/frontendOpenOrders.js");
const fundingHistory_js_1 = require("./_methods/fundingHistory.js");
const gossipRootIps_js_1 = require("./_methods/gossipRootIps.js");
const historicalOrders_js_1 = require("./_methods/historicalOrders.js");
const isVip_js_1 = require("./_methods/isVip.js");
const l2Book_js_1 = require("./_methods/l2Book.js");
const leadingVaults_js_1 = require("./_methods/leadingVaults.js");
const legalCheck_js_1 = require("./_methods/legalCheck.js");
const liquidatable_js_1 = require("./_methods/liquidatable.js");
const marginTable_js_1 = require("./_methods/marginTable.js");
const maxBuilderFee_js_1 = require("./_methods/maxBuilderFee.js");
const maxMarketOrderNtls_js_1 = require("./_methods/maxMarketOrderNtls.js");
const meta_js_1 = require("./_methods/meta.js");
const metaAndAssetCtxs_js_1 = require("./_methods/metaAndAssetCtxs.js");
const openOrders_js_1 = require("./_methods/openOrders.js");
const orderStatus_js_1 = require("./_methods/orderStatus.js");
const perpDeployAuctionStatus_js_1 = require("./_methods/perpDeployAuctionStatus.js");
const perpDexLimits_js_1 = require("./_methods/perpDexLimits.js");
const perpDexs_js_1 = require("./_methods/perpDexs.js");
const perpDexStatus_js_1 = require("./_methods/perpDexStatus.js");
const perpsAtOpenInterestCap_js_1 = require("./_methods/perpsAtOpenInterestCap.js");
const portfolio_js_1 = require("./_methods/portfolio.js");
const predictedFundings_js_1 = require("./_methods/predictedFundings.js");
const preTransferCheck_js_1 = require("./_methods/preTransferCheck.js");
const recentTrades_js_1 = require("./_methods/recentTrades.js");
const referral_js_1 = require("./_methods/referral.js");
const spotClearinghouseState_js_1 = require("./_methods/spotClearinghouseState.js");
const spotDeployState_js_1 = require("./_methods/spotDeployState.js");
const spotMeta_js_1 = require("./_methods/spotMeta.js");
const spotMetaAndAssetCtxs_js_1 = require("./_methods/spotMetaAndAssetCtxs.js");
const spotPairDeployAuctionStatus_js_1 = require("./_methods/spotPairDeployAuctionStatus.js");
const subAccounts_js_1 = require("./_methods/subAccounts.js");
const subAccounts2_js_1 = require("./_methods/subAccounts2.js");
const tokenDetails_js_1 = require("./_methods/tokenDetails.js");
const twapHistory_js_1 = require("./_methods/twapHistory.js");
const txDetails_js_1 = require("./_methods/txDetails.js");
const userDetails_js_1 = require("./_methods/userDetails.js");
const userDexAbstraction_js_1 = require("./_methods/userDexAbstraction.js");
const userFees_js_1 = require("./_methods/userFees.js");
const userFills_js_1 = require("./_methods/userFills.js");
const userFillsByTime_js_1 = require("./_methods/userFillsByTime.js");
const userFunding_js_1 = require("./_methods/userFunding.js");
const userNonFundingLedgerUpdates_js_1 = require("./_methods/userNonFundingLedgerUpdates.js");
const userRateLimit_js_1 = require("./_methods/userRateLimit.js");
const userRole_js_1 = require("./_methods/userRole.js");
const userToMultiSigSigners_js_1 = require("./_methods/userToMultiSigSigners.js");
const userTwapSliceFills_js_1 = require("./_methods/userTwapSliceFills.js");
const userTwapSliceFillsByTime_js_1 = require("./_methods/userTwapSliceFillsByTime.js");
const userVaultEquities_js_1 = require("./_methods/userVaultEquities.js");
const validatorL1Votes_js_1 = require("./_methods/validatorL1Votes.js");
const validatorSummaries_js_1 = require("./_methods/validatorSummaries.js");
const vaultDetails_js_1 = require("./_methods/vaultDetails.js");
const vaultSummaries_js_1 = require("./_methods/vaultSummaries.js");
const webData2_js_1 = require("./_methods/webData2.js");
// =============================================================
// Client
// =============================================================
/**
 * Read-only access to market data, user state, and other public information.
 *
 * Corresponds to the {@link https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/info-endpoint | Info endpoint}.
 */
class InfoClient {
    config_;
    /**
     * Creates an instance of the InfoClient.
     *
     * @param config - Configuration for Info API requests. See {@link InfoConfig}.
     *
     * @example
     * ```ts
     * import * as hl from "@nktkas/hyperliquid";
     *
     * const transport = new hl.HttpTransport(); // or `WebSocketTransport`
     *
     * const infoClient = new hl.InfoClient({ transport });
     * ```
     */
    constructor(config) {
        this.config_ = config;
    }
    /** @see {@link activeAssetData} */
    activeAssetData(params, signal) {
        return (0, activeAssetData_js_1.activeAssetData)(this.config_, params, signal);
    }
    /** @see {@link alignedQuoteTokenInfo} */
    alignedQuoteTokenInfo(params, signal) {
        return (0, alignedQuoteTokenInfo_js_1.alignedQuoteTokenInfo)(this.config_, params, signal);
    }
    allMids(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return (0, allMids_js_1.allMids)(this.config_, params, signal);
    }
    /** @see {@link allPerpMetas} */
    allPerpMetas(signal) {
        return (0, allPerpMetas_js_1.allPerpMetas)(this.config_, signal);
    }
    /** @see {@link blockDetails} */
    blockDetails(params, signal) {
        return (0, blockDetails_js_1.blockDetails)(this.config_, params, signal);
    }
    /** @see {@link candleSnapshot} */
    candleSnapshot(params, signal) {
        return (0, candleSnapshot_js_1.candleSnapshot)(this.config_, params, signal);
    }
    /** @see {@link clearinghouseState} */
    clearinghouseState(params, signal) {
        return (0, clearinghouseState_js_1.clearinghouseState)(this.config_, params, signal);
    }
    /** @see {@link delegations} */
    delegations(params, signal) {
        return (0, delegations_js_1.delegations)(this.config_, params, signal);
    }
    /** @see {@link delegatorHistory} */
    delegatorHistory(params, signal) {
        return (0, delegatorHistory_js_1.delegatorHistory)(this.config_, params, signal);
    }
    /** @see {@link delegatorRewards} */
    delegatorRewards(params, signal) {
        return (0, delegatorRewards_js_1.delegatorRewards)(this.config_, params, signal);
    }
    /** @see {@link delegatorSummary} */
    delegatorSummary(params, signal) {
        return (0, delegatorSummary_js_1.delegatorSummary)(this.config_, params, signal);
    }
    /** @see {@link exchangeStatus} */
    exchangeStatus(signal) {
        return (0, exchangeStatus_js_1.exchangeStatus)(this.config_, signal);
    }
    /** @see {@link extraAgents} */
    extraAgents(params, signal) {
        return (0, extraAgents_js_1.extraAgents)(this.config_, params, signal);
    }
    /** @see {@link frontendOpenOrders} */
    frontendOpenOrders(params, signal) {
        return (0, frontendOpenOrders_js_1.frontendOpenOrders)(this.config_, params, signal);
    }
    /** @see {@link fundingHistory} */
    fundingHistory(params, signal) {
        return (0, fundingHistory_js_1.fundingHistory)(this.config_, params, signal);
    }
    /** @see {@link gossipRootIps} */
    gossipRootIps(signal) {
        return (0, gossipRootIps_js_1.gossipRootIps)(this.config_, signal);
    }
    /** @see {@link historicalOrders} */
    historicalOrders(params, signal) {
        return (0, historicalOrders_js_1.historicalOrders)(this.config_, params, signal);
    }
    /** @see {@link isVip} */
    isVip(params, signal) {
        return (0, isVip_js_1.isVip)(this.config_, params, signal);
    }
    /** @see {@link l2Book} */
    l2Book(params, signal) {
        return (0, l2Book_js_1.l2Book)(this.config_, params, signal);
    }
    /** @see {@link leadingVaults} */
    leadingVaults(params, signal) {
        return (0, leadingVaults_js_1.leadingVaults)(this.config_, params, signal);
    }
    /** @see {@link legalCheck} */
    legalCheck(params, signal) {
        return (0, legalCheck_js_1.legalCheck)(this.config_, params, signal);
    }
    /** @see {@link liquidatable} */
    liquidatable(signal) {
        return (0, liquidatable_js_1.liquidatable)(this.config_, signal);
    }
    /** @see {@link marginTable} */
    marginTable(params, signal) {
        return (0, marginTable_js_1.marginTable)(this.config_, params, signal);
    }
    /** @see {@link maxBuilderFee} */
    maxBuilderFee(params, signal) {
        return (0, maxBuilderFee_js_1.maxBuilderFee)(this.config_, params, signal);
    }
    /** @see {@link maxMarketOrderNtls} */
    maxMarketOrderNtls(signal) {
        return (0, maxMarketOrderNtls_js_1.maxMarketOrderNtls)(this.config_, signal);
    }
    meta(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return (0, meta_js_1.meta)(this.config_, params, signal);
    }
    metaAndAssetCtxs(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return (0, metaAndAssetCtxs_js_1.metaAndAssetCtxs)(this.config_, params, signal);
    }
    /** @see {@link openOrders} */
    openOrders(params, signal) {
        return (0, openOrders_js_1.openOrders)(this.config_, params, signal);
    }
    /** @see {@link orderStatus} */
    orderStatus(params, signal) {
        return (0, orderStatus_js_1.orderStatus)(this.config_, params, signal);
    }
    /** @see {@link perpDeployAuctionStatus} */
    perpDeployAuctionStatus(signal) {
        return (0, perpDeployAuctionStatus_js_1.perpDeployAuctionStatus)(this.config_, signal);
    }
    /** @see {@link perpDexLimits} */
    perpDexLimits(params, signal) {
        return (0, perpDexLimits_js_1.perpDexLimits)(this.config_, params, signal);
    }
    /** @see {@link perpDexs} */
    perpDexs(signal) {
        return (0, perpDexs_js_1.perpDexs)(this.config_, signal);
    }
    /** @see {@link perpDexStatus} */
    perpDexStatus(params, signal) {
        return (0, perpDexStatus_js_1.perpDexStatus)(this.config_, params, signal);
    }
    perpsAtOpenInterestCap(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return (0, perpsAtOpenInterestCap_js_1.perpsAtOpenInterestCap)(this.config_, params, signal);
    }
    /** @see {@link portfolio} */
    portfolio(params, signal) {
        return (0, portfolio_js_1.portfolio)(this.config_, params, signal);
    }
    /** @see {@link predictedFundings} */
    predictedFundings(signal) {
        return (0, predictedFundings_js_1.predictedFundings)(this.config_, signal);
    }
    /** @see {@link preTransferCheck} */
    preTransferCheck(params, signal) {
        return (0, preTransferCheck_js_1.preTransferCheck)(this.config_, params, signal);
    }
    /** @see {@link recentTrades} */
    recentTrades(params, signal) {
        return (0, recentTrades_js_1.recentTrades)(this.config_, params, signal);
    }
    /** @see {@link referral} */
    referral(params, signal) {
        return (0, referral_js_1.referral)(this.config_, params, signal);
    }
    /** @see {@link spotClearinghouseState} */
    spotClearinghouseState(params, signal) {
        return (0, spotClearinghouseState_js_1.spotClearinghouseState)(this.config_, params, signal);
    }
    /** @see {@link spotDeployState} */
    spotDeployState(params, signal) {
        return (0, spotDeployState_js_1.spotDeployState)(this.config_, params, signal);
    }
    /** @see {@link spotMeta} */
    spotMeta(signal) {
        return (0, spotMeta_js_1.spotMeta)(this.config_, signal);
    }
    /** @see {@link spotMetaAndAssetCtxs} */
    spotMetaAndAssetCtxs(signal) {
        return (0, spotMetaAndAssetCtxs_js_1.spotMetaAndAssetCtxs)(this.config_, signal);
    }
    /** @see {@link spotPairDeployAuctionStatus} */
    spotPairDeployAuctionStatus(signal) {
        return (0, spotPairDeployAuctionStatus_js_1.spotPairDeployAuctionStatus)(this.config_, signal);
    }
    /** @see {@link subAccounts} */
    subAccounts(params, signal) {
        return (0, subAccounts_js_1.subAccounts)(this.config_, params, signal);
    }
    /** @see {@link subAccounts2} */
    subAccounts2(params, signal) {
        return (0, subAccounts2_js_1.subAccounts2)(this.config_, params, signal);
    }
    /** @see {@link tokenDetails} */
    tokenDetails(params, signal) {
        return (0, tokenDetails_js_1.tokenDetails)(this.config_, params, signal);
    }
    /** @see {@link twapHistory} */
    twapHistory(params, signal) {
        return (0, twapHistory_js_1.twapHistory)(this.config_, params, signal);
    }
    /** @see {@link txDetails} */
    txDetails(params, signal) {
        return (0, txDetails_js_1.txDetails)(this.config_, params, signal);
    }
    /** @see {@link userDetails} */
    userDetails(params, signal) {
        return (0, userDetails_js_1.userDetails)(this.config_, params, signal);
    }
    /** @see {@link userDexAbstraction} */
    userDexAbstraction(params, signal) {
        return (0, userDexAbstraction_js_1.userDexAbstraction)(this.config_, params, signal);
    }
    /** @see {@link userFees} */
    userFees(params, signal) {
        return (0, userFees_js_1.userFees)(this.config_, params, signal);
    }
    /** @see {@link userFills} */
    userFills(params, signal) {
        return (0, userFills_js_1.userFills)(this.config_, params, signal);
    }
    /** @see {@link userFillsByTime} */
    userFillsByTime(params, signal) {
        return (0, userFillsByTime_js_1.userFillsByTime)(this.config_, params, signal);
    }
    /** @see {@link userFunding} */
    userFunding(params, signal) {
        return (0, userFunding_js_1.userFunding)(this.config_, params, signal);
    }
    /** @see {@link userNonFundingLedgerUpdates} */
    userNonFundingLedgerUpdates(params, signal) {
        return (0, userNonFundingLedgerUpdates_js_1.userNonFundingLedgerUpdates)(this.config_, params, signal);
    }
    /** @see {@link userRateLimit} */
    userRateLimit(params, signal) {
        return (0, userRateLimit_js_1.userRateLimit)(this.config_, params, signal);
    }
    /** @see {@link userRole} */
    userRole(params, signal) {
        return (0, userRole_js_1.userRole)(this.config_, params, signal);
    }
    /** @see {@link userToMultiSigSigners} */
    userToMultiSigSigners(params, signal) {
        return (0, userToMultiSigSigners_js_1.userToMultiSigSigners)(this.config_, params, signal);
    }
    /** @see {@link userTwapSliceFills} */
    userTwapSliceFills(params, signal) {
        return (0, userTwapSliceFills_js_1.userTwapSliceFills)(this.config_, params, signal);
    }
    /** @see {@link userTwapSliceFillsByTime} */
    userTwapSliceFillsByTime(params, signal) {
        return (0, userTwapSliceFillsByTime_js_1.userTwapSliceFillsByTime)(this.config_, params, signal);
    }
    /** @see {@link userVaultEquities} */
    userVaultEquities(params, signal) {
        return (0, userVaultEquities_js_1.userVaultEquities)(this.config_, params, signal);
    }
    /** @see {@link validatorL1Votes} */
    validatorL1Votes(signal) {
        return (0, validatorL1Votes_js_1.validatorL1Votes)(this.config_, signal);
    }
    /** @see {@link validatorSummaries} */
    validatorSummaries(signal) {
        return (0, validatorSummaries_js_1.validatorSummaries)(this.config_, signal);
    }
    /** @see {@link vaultDetails} */
    vaultDetails(params, signal) {
        return (0, vaultDetails_js_1.vaultDetails)(this.config_, params, signal);
    }
    /** @see {@link vaultSummaries} */
    vaultSummaries(signal) {
        return (0, vaultSummaries_js_1.vaultSummaries)(this.config_, signal);
    }
    /** @see {@link webData2} */
    webData2(params, signal) {
        return (0, webData2_js_1.webData2)(this.config_, params, signal);
    }
}
exports.InfoClient = InfoClient;
