// @nktkas/hyperliquid/esm/src/api/info/client.js
// =============================================================
// Methods Imports
// =============================================================
import { activeAssetData, } from "./_methods/activeAssetData.js";
import { alignedQuoteTokenInfo, } from "./_methods/alignedQuoteTokenInfo.js";
import { allMids } from "./_methods/allMids.js";
import { allPerpMetas } from "./_methods/allPerpMetas.js";
import { blockDetails } from "./_methods/blockDetails.js";
import { candleSnapshot, } from "./_methods/candleSnapshot.js";
import { clearinghouseState, } from "./_methods/clearinghouseState.js";
import { delegations } from "./_methods/delegations.js";
import { delegatorHistory, } from "./_methods/delegatorHistory.js";
import { delegatorRewards, } from "./_methods/delegatorRewards.js";
import { delegatorSummary, } from "./_methods/delegatorSummary.js";
import { exchangeStatus } from "./_methods/exchangeStatus.js";
import { extraAgents } from "./_methods/extraAgents.js";
import { frontendOpenOrders, } from "./_methods/frontendOpenOrders.js";
import { fundingHistory, } from "./_methods/fundingHistory.js";
import { gossipRootIps } from "./_methods/gossipRootIps.js";
import { historicalOrders, } from "./_methods/historicalOrders.js";
import { isVip } from "./_methods/isVip.js";
import { l2Book } from "./_methods/l2Book.js";
import { leadingVaults } from "./_methods/leadingVaults.js";
import { legalCheck } from "./_methods/legalCheck.js";
import { liquidatable } from "./_methods/liquidatable.js";
import { marginTable } from "./_methods/marginTable.js";
import { maxBuilderFee } from "./_methods/maxBuilderFee.js";
import { maxMarketOrderNtls } from "./_methods/maxMarketOrderNtls.js";
import { meta } from "./_methods/meta.js";
import { metaAndAssetCtxs, } from "./_methods/metaAndAssetCtxs.js";
import { openOrders } from "./_methods/openOrders.js";
import { orderStatus } from "./_methods/orderStatus.js";
import { perpDeployAuctionStatus } from "./_methods/perpDeployAuctionStatus.js";
import { perpDexLimits } from "./_methods/perpDexLimits.js";
import { perpDexs } from "./_methods/perpDexs.js";
import { perpDexStatus } from "./_methods/perpDexStatus.js";
import { perpsAtOpenInterestCap, } from "./_methods/perpsAtOpenInterestCap.js";
import { portfolio } from "./_methods/portfolio.js";
import { predictedFundings } from "./_methods/predictedFundings.js";
import { preTransferCheck, } from "./_methods/preTransferCheck.js";
import { recentTrades } from "./_methods/recentTrades.js";
import { referral } from "./_methods/referral.js";
import { spotClearinghouseState, } from "./_methods/spotClearinghouseState.js";
import { spotDeployState, } from "./_methods/spotDeployState.js";
import { spotMeta } from "./_methods/spotMeta.js";
import { spotMetaAndAssetCtxs } from "./_methods/spotMetaAndAssetCtxs.js";
import { spotPairDeployAuctionStatus, } from "./_methods/spotPairDeployAuctionStatus.js";
import { subAccounts } from "./_methods/subAccounts.js";
import { subAccounts2 } from "./_methods/subAccounts2.js";
import { tokenDetails } from "./_methods/tokenDetails.js";
import { twapHistory } from "./_methods/twapHistory.js";
import { txDetails } from "./_methods/txDetails.js";
import { userDetails } from "./_methods/userDetails.js";
import { userDexAbstraction, } from "./_methods/userDexAbstraction.js";
import { userFees } from "./_methods/userFees.js";
import { userFills } from "./_methods/userFills.js";
import { userFillsByTime, } from "./_methods/userFillsByTime.js";
import { userFunding } from "./_methods/userFunding.js";
import { userNonFundingLedgerUpdates, } from "./_methods/userNonFundingLedgerUpdates.js";
import { userRateLimit } from "./_methods/userRateLimit.js";
import { userRole } from "./_methods/userRole.js";
import { userToMultiSigSigners, } from "./_methods/userToMultiSigSigners.js";
import { userTwapSliceFills, } from "./_methods/userTwapSliceFills.js";
import { userTwapSliceFillsByTime, } from "./_methods/userTwapSliceFillsByTime.js";
import { userVaultEquities, } from "./_methods/userVaultEquities.js";
import { validatorL1Votes } from "./_methods/validatorL1Votes.js";
import { validatorSummaries } from "./_methods/validatorSummaries.js";
import { vaultDetails } from "./_methods/vaultDetails.js";
import { vaultSummaries } from "./_methods/vaultSummaries.js";
import { webData2 } from "./_methods/webData2.js";
// =============================================================
// Client
// =============================================================
/**
 * Read-only access to market data, user state, and other public information.
 *
 * Corresponds to the {@link https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/info-endpoint | Info endpoint}.
 */
export class InfoClient {
    config_;
    /**
     * Creates an instance of the InfoClient.
     *
     * @param config - Configuration for Info API requests. See {@link InfoConfig}.
     *
     * @example
     * ```ts
     * import * as hl from "@nktkas/hyperliquid";
     *
     * const transport = new hl.HttpTransport(); // or `WebSocketTransport`
     *
     * const infoClient = new hl.InfoClient({ transport });
     * ```
     */
    constructor(config) {
        this.config_ = config;
    }
    /** @see {@link activeAssetData} */
    activeAssetData(params, signal) {
        return activeAssetData(this.config_, params, signal);
    }
    /** @see {@link alignedQuoteTokenInfo} */
    alignedQuoteTokenInfo(params, signal) {
        return alignedQuoteTokenInfo(this.config_, params, signal);
    }
    allMids(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return allMids(this.config_, params, signal);
    }
    /** @see {@link allPerpMetas} */
    allPerpMetas(signal) {
        return allPerpMetas(this.config_, signal);
    }
    /** @see {@link blockDetails} */
    blockDetails(params, signal) {
        return blockDetails(this.config_, params, signal);
    }
    /** @see {@link candleSnapshot} */
    candleSnapshot(params, signal) {
        return candleSnapshot(this.config_, params, signal);
    }
    /** @see {@link clearinghouseState} */
    clearinghouseState(params, signal) {
        return clearinghouseState(this.config_, params, signal);
    }
    /** @see {@link delegations} */
    delegations(params, signal) {
        return delegations(this.config_, params, signal);
    }
    /** @see {@link delegatorHistory} */
    delegatorHistory(params, signal) {
        return delegatorHistory(this.config_, params, signal);
    }
    /** @see {@link delegatorRewards} */
    delegatorRewards(params, signal) {
        return delegatorRewards(this.config_, params, signal);
    }
    /** @see {@link delegatorSummary} */
    delegatorSummary(params, signal) {
        return delegatorSummary(this.config_, params, signal);
    }
    /** @see {@link exchangeStatus} */
    exchangeStatus(signal) {
        return exchangeStatus(this.config_, signal);
    }
    /** @see {@link extraAgents} */
    extraAgents(params, signal) {
        return extraAgents(this.config_, params, signal);
    }
    /** @see {@link frontendOpenOrders} */
    frontendOpenOrders(params, signal) {
        return frontendOpenOrders(this.config_, params, signal);
    }
    /** @see {@link fundingHistory} */
    fundingHistory(params, signal) {
        return fundingHistory(this.config_, params, signal);
    }
    /** @see {@link gossipRootIps} */
    gossipRootIps(signal) {
        return gossipRootIps(this.config_, signal);
    }
    /** @see {@link historicalOrders} */
    historicalOrders(params, signal) {
        return historicalOrders(this.config_, params, signal);
    }
    /** @see {@link isVip} */
    isVip(params, signal) {
        return isVip(this.config_, params, signal);
    }
    /** @see {@link l2Book} */
    l2Book(params, signal) {
        return l2Book(this.config_, params, signal);
    }
    /** @see {@link leadingVaults} */
    leadingVaults(params, signal) {
        return leadingVaults(this.config_, params, signal);
    }
    /** @see {@link legalCheck} */
    legalCheck(params, signal) {
        return legalCheck(this.config_, params, signal);
    }
    /** @see {@link liquidatable} */
    liquidatable(signal) {
        return liquidatable(this.config_, signal);
    }
    /** @see {@link marginTable} */
    marginTable(params, signal) {
        return marginTable(this.config_, params, signal);
    }
    /** @see {@link maxBuilderFee} */
    maxBuilderFee(params, signal) {
        return maxBuilderFee(this.config_, params, signal);
    }
    /** @see {@link maxMarketOrderNtls} */
    maxMarketOrderNtls(signal) {
        return maxMarketOrderNtls(this.config_, signal);
    }
    meta(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return meta(this.config_, params, signal);
    }
    metaAndAssetCtxs(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return metaAndAssetCtxs(this.config_, params, signal);
    }
    /** @see {@link openOrders} */
    openOrders(params, signal) {
        return openOrders(this.config_, params, signal);
    }
    /** @see {@link orderStatus} */
    orderStatus(params, signal) {
        return orderStatus(this.config_, params, signal);
    }
    /** @see {@link perpDeployAuctionStatus} */
    perpDeployAuctionStatus(signal) {
        return perpDeployAuctionStatus(this.config_, signal);
    }
    /** @see {@link perpDexLimits} */
    perpDexLimits(params, signal) {
        return perpDexLimits(this.config_, params, signal);
    }
    /** @see {@link perpDexs} */
    perpDexs(signal) {
        return perpDexs(this.config_, signal);
    }
    /** @see {@link perpDexStatus} */
    perpDexStatus(params, signal) {
        return perpDexStatus(this.config_, params, signal);
    }
    perpsAtOpenInterestCap(paramsOrSignal, maybeSignal) {
        const params = paramsOrSignal instanceof AbortSignal ? {} : paramsOrSignal;
        const signal = paramsOrSignal instanceof AbortSignal ? paramsOrSignal : maybeSignal;
        return perpsAtOpenInterestCap(this.config_, params, signal);
    }
    /** @see {@link portfolio} */
    portfolio(params, signal) {
        return portfolio(this.config_, params, signal);
    }
    /** @see {@link predictedFundings} */
    predictedFundings(signal) {
        return predictedFundings(this.config_, signal);
    }
    /** @see {@link preTransferCheck} */
    preTransferCheck(params, signal) {
        return preTransferCheck(this.config_, params, signal);
    }
    /** @see {@link recentTrades} */
    recentTrades(params, signal) {
        return recentTrades(this.config_, params, signal);
    }
    /** @see {@link referral} */
    referral(params, signal) {
        return referral(this.config_, params, signal);
    }
    /** @see {@link spotClearinghouseState} */
    spotClearinghouseState(params, signal) {
        return spotClearinghouseState(this.config_, params, signal);
    }
    /** @see {@link spotDeployState} */
    spotDeployState(params, signal) {
        return spotDeployState(this.config_, params, signal);
    }
    /** @see {@link spotMeta} */
    spotMeta(signal) {
        return spotMeta(this.config_, signal);
    }
    /** @see {@link spotMetaAndAssetCtxs} */
    spotMetaAndAssetCtxs(signal) {
        return spotMetaAndAssetCtxs(this.config_, signal);
    }
    /** @see {@link spotPairDeployAuctionStatus} */
    spotPairDeployAuctionStatus(signal) {
        return spotPairDeployAuctionStatus(this.config_, signal);
    }
    /** @see {@link subAccounts} */
    subAccounts(params, signal) {
        return subAccounts(this.config_, params, signal);
    }
    /** @see {@link subAccounts2} */
    subAccounts2(params, signal) {
        return subAccounts2(this.config_, params, signal);
    }
    /** @see {@link tokenDetails} */
    tokenDetails(params, signal) {
        return tokenDetails(this.config_, params, signal);
    }
    /** @see {@link twapHistory} */
    twapHistory(params, signal) {
        return twapHistory(this.config_, params, signal);
    }
    /** @see {@link txDetails} */
    txDetails(params, signal) {
        return txDetails(this.config_, params, signal);
    }
    /** @see {@link userDetails} */
    userDetails(params, signal) {
        return userDetails(this.config_, params, signal);
    }
    /** @see {@link userDexAbstraction} */
    userDexAbstraction(params, signal) {
        return userDexAbstraction(this.config_, params, signal);
    }
    /** @see {@link userFees} */
    userFees(params, signal) {
        return userFees(this.config_, params, signal);
    }
    /** @see {@link userFills} */
    userFills(params, signal) {
        return userFills(this.config_, params, signal);
    }
    /** @see {@link userFillsByTime} */
    userFillsByTime(params, signal) {
        return userFillsByTime(this.config_, params, signal);
    }
    /** @see {@link userFunding} */
    userFunding(params, signal) {
        return userFunding(this.config_, params, signal);
    }
    /** @see {@link userNonFundingLedgerUpdates} */
    userNonFundingLedgerUpdates(params, signal) {
        return userNonFundingLedgerUpdates(this.config_, params, signal);
    }
    /** @see {@link userRateLimit} */
    userRateLimit(params, signal) {
        return userRateLimit(this.config_, params, signal);
    }
    /** @see {@link userRole} */
    userRole(params, signal) {
        return userRole(this.config_, params, signal);
    }
    /** @see {@link userToMultiSigSigners} */
    userToMultiSigSigners(params, signal) {
        return userToMultiSigSigners(this.config_, params, signal);
    }
    /** @see {@link userTwapSliceFills} */
    userTwapSliceFills(params, signal) {
        return userTwapSliceFills(this.config_, params, signal);
    }
    /** @see {@link userTwapSliceFillsByTime} */
    userTwapSliceFillsByTime(params, signal) {
        return userTwapSliceFillsByTime(this.config_, params, signal);
    }
    /** @see {@link userVaultEquities} */
    userVaultEquities(params, signal) {
        return userVaultEquities(this.config_, params, signal);
    }
    /** @see {@link validatorL1Votes} */
    validatorL1Votes(signal) {
        return validatorL1Votes(this.config_, signal);
    }
    /** @see {@link validatorSummaries} */
    validatorSummaries(signal) {
        return validatorSummaries(this.config_, signal);
    }
    /** @see {@link vaultDetails} */
    vaultDetails(params, signal) {
        return vaultDetails(this.config_, params, signal);
    }
    /** @see {@link vaultSummaries} */
    vaultSummaries(signal) {
        return vaultSummaries(this.config_, signal);
    }
    /** @see {@link webData2} */
    webData2(params, signal) {
        return webData2(this.config_, params, signal);
    }
}
