// @nktkas/hyperliquid/script/src/api/subscription/_methods/bbo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BboEvent = exports.BboRequest = void 0;
exports.bbo = bbo;
const v = __importStar(require("valibot"));
// ============================================================
// API Schemas
// ============================================================
const _schemas_js_1 = require("../../_schemas.js");
/** Subscription to best bid and offer events for a specific asset. */
exports.BboRequest = (() => {
    return v.pipe(v.object({
        /** Type of subscription. */
        type: v.pipe(v.literal("bbo"), v.description("Type of subscription.")),
        /** Asset symbol (e.g., BTC). */
        coin: v.pipe(v.string(), v.description("Asset symbol (e.g., BTC).")),
    }), v.description("Subscription to best bid and offer events for a specific asset."));
})();
const L2BookLevelSchema = /* @__PURE__ */ (() => {
    return v.pipe(v.object({
        /** Price. */
        px: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Price.")),
        /** Total size. */
        sz: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Total size.")),
        /** Number of individual orders. */
        n: v.pipe(_schemas_js_1.UnsignedInteger, v.description("Number of individual orders.")),
    }), v.description("L2 order book level."));
})();
/** Event of best bid and offer. */
exports.BboEvent = (() => {
    return v.pipe(v.object({
        /** Asset symbol (e.g., BTC). */
        coin: v.pipe(v.string(), v.description("Asset symbol (e.g., BTC).")),
        /** Time of the BBO update (in ms since epoch). */
        time: v.pipe(_schemas_js_1.UnsignedInteger, v.description("Time of the BBO update (in ms since epoch).")),
        /** Best bid and offer tuple [bid, offer], either can be undefined if unavailable. */
        bbo: v.pipe(v.tuple([L2BookLevelSchema, L2BookLevelSchema]), v.description("Best bid and offer tuple [bid, offer], either can be undefined if unavailable.")),
    }), v.description("Event of best bid and offer."));
})();
/**
 * Subscribe to best bid and offer updates for a specific asset.
 *
 * @param config - General configuration for Subscription API subscriptions.
 * @param params - Parameters specific to the API subscription.
 * @param listener - A callback function to be called when the event is received.
 *
 * @returns A request-promise that resolves with a {@link ISubscription} object to manage the subscription lifecycle.
 *
 * @throws {ValiError} When the request parameters fail validation (before sending).
 * @throws {TransportError} When the transport layer throws an error.
 *
 * @example
 * ```ts
 * import { WebSocketTransport } from "@nktkas/hyperliquid";
 * import { bbo } from "@nktkas/hyperliquid/api/subscription";
 *
 * const transport = new WebSocketTransport(); // only `WebSocketTransport`
 *
 * const sub = await bbo(
 *   { transport },
 *   { coin: "ETH" },
 *   (data) => console.log(data),
 * );
 * ```
 *
 * @see https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/websocket/subscriptions
 */
function bbo(config, params, listener) {
    const payload = v.parse(exports.BboRequest, { type: "bbo", ...params });
    return config.transport.subscribe(payload.type, payload, (e) => {
        if (e.detail.coin === payload.coin) {
            listener(e.detail);
        }
    });
}
