// @nktkas/hyperliquid/script/src/api/subscription/_methods/candle.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CandleEvent = exports.CandleRequest = void 0;
exports.candle = candle;
const v = __importStar(require("valibot"));
// ============================================================
// API Schemas
// ============================================================
const _schemas_js_1 = require("../../_schemas.js");
/** Subscription to candlestick events for a specific asset and time interval. */
exports.CandleRequest = (() => {
    return v.pipe(v.object({
        /** Type of subscription. */
        type: v.pipe(v.literal("candle"), v.description("Type of subscription.")),
        /** Asset symbol (e.g., BTC). */
        coin: v.pipe(v.string(), v.description("Asset symbol (e.g., BTC).")),
        /** Time interval. */
        interval: v.pipe(v.picklist(["1m", "3m", "5m", "15m", "30m", "1h", "2h", "4h", "8h", "12h", "1d", "3d", "1w", "1M"]), v.description("Time interval.")),
    }), v.description("Subscription to candlestick events for a specific asset and time interval."));
})();
/** Event of candlestick data point. */
exports.CandleEvent = (() => {
    return v.pipe(v.object({
        /** Opening timestamp (ms since epoch). */
        t: v.pipe(_schemas_js_1.UnsignedInteger, v.description("Opening timestamp (ms since epoch).")),
        /** Closing timestamp (ms since epoch). */
        T: v.pipe(_schemas_js_1.UnsignedInteger, v.description("Closing timestamp (ms since epoch).")),
        /** Asset symbol. */
        s: v.pipe(v.string(), v.description("Asset symbol.")),
        /** Time interval. */
        i: v.pipe(v.picklist(["1m", "3m", "5m", "15m", "30m", "1h", "2h", "4h", "8h", "12h", "1d", "3d", "1w", "1M"]), v.description("Time interval.")),
        /** Opening price. */
        o: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Opening price.")),
        /** Closing price. */
        c: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Closing price.")),
        /** Highest price. */
        h: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Highest price.")),
        /** Lowest price. */
        l: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Lowest price.")),
        /** Total volume traded in base currency. */
        v: v.pipe(_schemas_js_1.UnsignedDecimal, v.description("Total volume traded in base currency.")),
        /** Number of trades executed. */
        n: v.pipe(_schemas_js_1.UnsignedInteger, v.description("Number of trades executed.")),
    }), v.description("Event of candlestick data point."));
})();
/**
 * Subscribe to candlestick data updates for a specific asset.
 *
 * @param config - General configuration for Subscription API subscriptions.
 * @param params - Parameters specific to the API subscription.
 * @param listener - A callback function to be called when the event is received.
 *
 * @returns A request-promise that resolves with a {@link ISubscription} object to manage the subscription lifecycle.
 *
 * @throws {ValiError} When the request parameters fail validation (before sending).
 * @throws {TransportError} When the transport layer throws an error.
 *
 * @example
 * ```ts
 * import { WebSocketTransport } from "@nktkas/hyperliquid";
 * import { candle } from "@nktkas/hyperliquid/api/subscription";
 *
 * const transport = new WebSocketTransport(); // only `WebSocketTransport`
 *
 * const sub = await candle(
 *   { transport },
 *   { coin: "ETH", interval: "1h" },
 *   (data) => console.log(data),
 * );
 * ```
 *
 * @see https://hyperliquid.gitbook.io/hyperliquid-docs/for-developers/api/websocket/subscriptions
 */
function candle(config, params, listener) {
    const payload = v.parse(exports.CandleRequest, { type: "candle", ...params });
    return config.transport.subscribe(payload.type, payload, (e) => {
        if (e.detail.s === payload.coin && e.detail.i === payload.interval) {
            listener(e.detail);
        }
    });
}
